### PR TITLE
Add ability to omit root element

### PR DESCRIPTION
**Pull request by jepp2078**
See https://github.com/einride/protobuf-avro-go/issues/67

### CODE REVIEW

**Comment by brunsgaard** [comment]
@ericwenn 👀

**Comment by ericwenn** [comment]
> @ericwenn 👀

Thanks for PR! First glance it looks good! 
Will take a closer look tomorrow!

File: encoding/protoavro/encode.go [hunk 1]
```diff
@@ -47,69 +47,72 @@ func messageJSON(message protoreflect.Message) (interface{}, error) {
...
 		}
 		value := message.Get(field)
-		jsonValue, err := fieldJSON(field, value)
+		jsonValue, err := o.fieldJSON(field, value)
 		if err != nil {
 			return nil, err
 		}
 		record[string(field.Name())] = jsonValue
 	}
+	if o.OmitRootElement && message.Descriptor().Index() == 0 {
```
Comment (ericwenn) [inline encoding/protoavro/encode.go]: Why is `.Descriptor().Index` necessary here?

File: encoding/protoavro/marshal.go [hunk 2]
```diff
@@ -13,7 +13,13 @@ import (
 // NewMarshaler returns a new marshaler that writes protobuf messages to writer in
 // Avro binary format.
 func NewMarshaler(descriptor protoreflect.MessageDescriptor, writer io.Writer) (*Marshaler, error) {
```
Comment (ericwenn) [inline encoding/protoavro/marshal.go:15]: nit: add to func comment that it uses default `SchemaOptions`

File: encoding/protoavro/marshal.go [hunk 3]
```diff
@@ -28,11 +34,12 @@ func NewMarshaler(descriptor protoreflect.MessageDescriptor, writer io.Writer) (
 	if err != nil {
 		return nil, fmt.Errorf("new ocf writer: %w", err)
 	}
-	return &Marshaler{w: w, desc: descriptor}, nil
+	return &Marshaler{w: w, desc: descriptor, SchemaOptions: o}, nil
 }
 
 // Marshaler encodes and writes Avro binary encoded messages.
 type Marshaler struct {
+	SchemaOptions
```
Comment (ericwenn) [inline encoding/protoavro/marshal.go:42]: Use a regular field for this, we don't want to export the `SchemaOptions` on an already created `Marshaler`.
Ie `opts SchemaOptions`

File: encoding/protoavro/schema.go [hunk 4]
```diff
@@ -8,17 +8,28 @@ import (
...
+	return SchemaOptions{}.newSchemaInferrer().inferMessageSchema(desc)
+}
+
+// InferSchema returns the Avro schema for the protobuf message descriptor.
+func (o SchemaOptions) InferSchema(desc protoreflect.MessageDescriptor) (avro.Schema, error) {
+	return o.newSchemaInferrer().inferMessageSchema(desc)
 }
 
 type schemaInferrer struct {
+	SchemaOptions
```
Comment (ericwenn) [inline encoding/protoavro/schema.go:27]: Same as above, add a regular field for `SchemaOptions`

File: encoding/protoavro/schema.go [hunk 5]
```diff
@@ -8,17 +8,28 @@ import (
 	"google.golang.org/protobuf/reflect/protoreflect"
 )
 
+// SchemaOptions contains configuration options for Avro schema inference.
+type SchemaOptions struct {
+	OmitRootElement bool
+}
+
 // InferSchema returns the Avro schema for the protobuf message descriptor.
 func InferSchema(desc protoreflect.MessageDescriptor) (avro.Schema, error) {
```
Comment (ericwenn) [inline encoding/protoavro/schema.go:17]: nit: add func comment that it uses the default options.

File: encoding/protoavro/schema.go [hunk 6]
```diff
@@ -8,17 +8,28 @@ import (
 	"google.golang.org/protobuf/reflect/protoreflect"
 )
 
+// SchemaOptions contains configuration options for Avro schema inference.
```
Comment (ericwenn) [inline encoding/protoavro/schema.go:11]: This is not only used for schema inference, so lets drop that from the comment.

File: encoding/protoavro/schema.go [hunk 7]
```diff
@@ -8,17 +8,28 @@ import (
...
+type SchemaOptions struct {
```
Comment (ericwenn) [inline encoding/protoavro/schema.go:12]: How about putting this in a separate file `options.go`?

File: encoding/protoavro/schema.go [hunk 8]
```diff
@@ -8,17 +8,28 @@ import (
...
+	OmitRootElement bool
```
Comment (ericwenn) [inline encoding/protoavro/schema.go:13]: Lets add a comment on this.

File: encoding/protoavro/unmarshal.go [hunk 9]
```diff
@@ -18,8 +18,19 @@ func NewUnmarshaler(reader io.Reader) (*Unmarshaler, error) {
...
+	r, err := goavro.NewOCFReader(reader)
+	if err != nil {
+		return nil, fmt.Errorf("new ocf writer: %w", err)
+	}
+	return &Unmarshaler{SchemaOptions: o, r: r}, nil
+}
+
 // Unmarshaler reads and decodes Avro binary encoded messages.
 type Unmarshaler struct {
+	SchemaOptions
```
Comment (ericwenn) [inline encoding/protoavro/unmarshal.go:33]: Same as above, regular field.

**Comment by ericwenn** [comment]
Looks good to me apart from some small changes! Nice job!

For CI you can run `make` locally until it passes (if `make` passes, so will CI).
Please also squash your commits (I think one commit will be fine) and prefix with `feat: [...]`

File: encoding/protoavro/encode.go (hunk 1 above)
Comment (jepp2078) [reply encoding/protoavro/encode.go]: Was removed in later commits

**Comment by jepp2078** [comment]
> Looks good to me apart from some small changes! Nice job!
> 
> For CI you can run `make` locally until it passes (if `make` passes, so will CI). Please also squash your commits (I think one commit will be fine) and prefix with `feat: [...]`

Fixed! :-)

**Comment by jepp2078** [comment]
@ericwenn can you give this a look when you have the chance? :eye: :eye:

**Comment by ericwenn** [comment]
Please squash all the commits and add `feat: [...]` prefix to the commit message.

**Comment by jepp2078** [comment]
> Please squash all the commits and add `feat: [...]` prefix to the commit message.

done

**Comment by ericwenn** [comment]
Thanks for the contribution @jepp2078

**Review by ericwenn — APPROVED** [review]

**Pull request merged**